File: core/encoding/cyrillic-to-latin/cs/CyrillicToLatinFallback.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

class CyrillicToLatinFallback : EncoderFallback
{
    private Dictionary<Char, String> table;

    public CyrillicToLatinFallback()
    {
        table = new Dictionary<Char, String>();
        // Define mappings.
        // Uppercase modern Cyrillic characters.
        table.Add('\u0410', "A");
        table.Add('\u0411', "B");
        table.Add('\u0412', "V");
        table.Add('\u0413', "G");
        table.Add('\u0414', "D");
        table.Add('\u0415', "E");
        table.Add('\u0416', "Zh");
        table.Add('\u0417', "Z");
        table.Add('\u0418', "I");
        table.Add('\u0419', "I");
        table.Add('\u041A', "K");
        table.Add('\u041B', "L");
        table.Add('\u041C', "M");
        table.Add('\u041D', "N");
        table.Add('\u041E', "O");
        table.Add('\u041F', "P");
        table.Add('\u0420', "R");
        table.Add('\u0421', "S");
        table.Add('\u0422', "T");
        table.Add('\u0423', "U");
        table.Add('\u0424', "F");
        table.Add('\u0425', "Kh");
        table.Add('\u0426', "Ts");
        table.Add('\u0427', "Ch");
        table.Add('\u0428', "Sh");
        table.Add('\u0429', "Shch");
        table.Add('\u042A', "'");    // Hard sign
        table.Add('\u042B', "Ye");
        table.Add('\u042C', "'");    // Soft sign
        table.Add('\u042D', "E");
        table.Add('\u042E', "Iu");
        table.Add('\u042F', "Ia");
        // Lowercase modern Cyrillic characters.
        table.Add('\u0430', "a");
        table.Add('\u0431', "b");
        table.Add('\u0432', "v");
        table.Add('\u0433', "g");
        table.Add('\u0434', "d");
        table.Add('\u0435', "e");
        table.Add('\u0436', "zh");
        table.Add('\u0437', "z");
        table.Add('\u0438', "i");
        table.Add('\u0439', "i");
        table.Add('\u043A', "k");
        table.Add('\u043B', "l");
        table.Add('\u043C', "m");
        table.Add('\u043D', "n");
        table.Add('\u043E', "o");
        table.Add('\u043F', "p");
        table.Add('\u0440', "r");
        table.Add('\u0441', "s");
        table.Add('\u0442', "t");
        table.Add('\u0443', "u");
        table.Add('\u0444', "f");
        table.Add('\u0445', "kh");
        table.Add('\u0446', "ts");
        table.Add('\u0447', "ch");
        table.Add('\u0448', "sh");
        table.Add('\u0449', "shch");
        table.Add('\u044A', "'");   // Hard sign
        table.Add('\u044B', "yi");
        table.Add('\u044C', "'");   // Soft sign
        table.Add('\u044D', "e");
        table.Add('\u044E', "iu");
        table.Add('\u044F', "ia");
    }

    public override EncoderFallbackBuffer CreateFallbackBuffer()
    {
        return new CyrillicToLatinFallbackBuffer(table);
    }

    public override int MaxCharCount
    {
        get { return 4; }                  // Maximum is "Shch" and "shch"
    }
}

public class CyrillicToLatinFallbackBuffer : EncoderFallbackBuffer
{
    private Dictionary<Char, String> table;
    private int bufferIndex;
    private string buffer;
    private int leftToReturn;

    internal CyrillicToLatinFallbackBuffer(Dictionary<Char, String> table)
    {
        this.table = table;
        this.bufferIndex = -1;
        this.leftToReturn = -1;
    }

    public override bool Fallback(char charUnknownHigh, char charUnknownLow, int index)
    {
        // There's no need to handle surrogates.
        return false;
    }

    public override bool Fallback(char charUnknown, int index)
    {
        if (charUnknown >= '\u0410' & charUnknown <= '\u044F')
        {
            buffer = table[charUnknown];
            leftToReturn = buffer.Length - 1;
            bufferIndex = -1;
            return true;
        }
        return false;
    }

    public override char GetNextChar()
    {
        char charToReturn;
        if (leftToReturn >= 0)
        {
            leftToReturn--;
            bufferIndex++;
            charToReturn = buffer[bufferIndex];
        }
        else
        {
            charToReturn = '\u0000';
        }
        return charToReturn;
    }

    public override bool MovePrevious()
    {
        if (bufferIndex > 0)
        {
            bufferIndex--;
            leftToReturn++;
            return true;
        }
        return false;
    }

    public override int Remaining
    {
        get { return leftToReturn; }
    }
}

File: core/encoding/cyrillic-to-latin/vb/CyrillicToLatinFallback.vb
Imports System.IO
Imports System.Text

Public Class CyrillicToLatinFallback : Inherits EncoderFallback

    Private table As Dictionary(Of Char, String)

    Public Sub New()
        table = New Dictionary(Of Char, String)
        ' Define mappings.
        ' Uppercase modern Cyrillic characters.
        table.Add(ChrW(&H0410), "A")
        table.Add(ChrW(&H0411), "B")
        table.Add(ChrW(&H0412), "V")
        table.Add(ChrW(&H0413), "G")
        table.Add(ChrW(&H0414), "D")
        table.Add(ChrW(&H0415), "E")
        table.Add(ChrW(&H0416), "Zh")
        table.Add(ChrW(&H0417), "Z")
        table.Add(ChrW(&H0418), "I")
        table.Add(ChrW(&H0419), "I")
        table.Add(ChrW(&H041A), "K")
        table.Add(ChrW(&H041B), "L")
        table.Add(ChrW(&H041C), "M")
        table.Add(ChrW(&H041D), "N")
        table.Add(ChrW(&H041E), "O")
        table.Add(ChrW(&H041F), "P")
        table.Add(ChrW(&H0420), "R")
        table.Add(ChrW(&H0421), "S")
        table.Add(ChrW(&H0422), "T")
        table.Add(ChrW(&H0423), "U")
        table.Add(ChrW(&H0424), "F")
        table.Add(ChrW(&H0425), "Kh")
        table.Add(ChrW(&H0426), "Ts")
        table.Add(ChrW(&H0427), "Ch")
        table.Add(ChrW(&H0428), "Sh")
        table.Add(ChrW(&H0429), "Shch")
        table.Add(ChrW(&H042A), "'")    ' Hard sign
        table.Add(ChrW(&H042B), "Ye")
        table.Add(ChrW(&H042C), "'")    ' Soft sign
        table.Add(ChrW(&H042D), "E")
        table.Add(ChrW(&H042E), "Iu")
        table.Add(ChrW(&H042F), "Ia")
        ' Lowercase modern Cyrillic characters.
        table.Add(ChrW(&H0430), "a")
        table.Add(ChrW(&H0431), "b")
        table.Add(ChrW(&H0432), "v")
        table.Add(ChrW(&H0433), "g")
        table.Add(ChrW(&H0434), "d")
        table.Add(ChrW(&H0435), "e")
        table.Add(ChrW(&H0436), "zh")
        table.Add(ChrW(&H0437), "z")
        table.Add(ChrW(&H0438), "i")
        table.Add(ChrW(&H0439), "i")
        table.Add(ChrW(&H043A), "k")
        table.Add(ChrW(&H043B), "l")
        table.Add(ChrW(&H043C), "m")
        table.Add(ChrW(&H043D), "n")
        table.Add(ChrW(&H043E), "o")
        table.Add(ChrW(&H043F), "p")
        table.Add(ChrW(&H0440), "r")
        table.Add(ChrW(&H0441), "s")
        table.Add(ChrW(&H0442), "t")
        table.Add(ChrW(&H0443), "u")
        table.Add(ChrW(&H0444), "f")
        table.Add(ChrW(&H0445), "kh")
        table.Add(ChrW(&H0446), "ts")
        table.Add(ChrW(&H0447), "ch")
        table.Add(ChrW(&H0448), "sh")
        table.Add(ChrW(&H0449), "shch")
        table.Add(ChrW(&H044A), "'")   ' Hard sign
        table.Add(ChrW(&H044B), "yi")
        table.Add(ChrW(&H044C), "'")   ' Soft sign
        table.Add(ChrW(&H044D), "e")
        table.Add(ChrW(&H044E), "iu")
        table.Add(ChrW(&H044F), "ia")
    End Sub

    Public Overrides Function CreateFallbackBuffer() As System.Text.EncoderFallbackBuffer
        Return New CyrillicToLatinFallbackBuffer(table)
    End Function

    Public Overrides ReadOnly Property MaxCharCount As Integer
        Get
            Return 4                  ' Maximum is "Shch" and "shch"
        End Get
    End Property
End Class

Public Class CyrillicToLatinFallbackBuffer : Inherits EncoderFallbackBuffer
    Private table As Dictionary(Of Char, String)
    Private bufferIndex As Integer
    Private buffer As String
    Private leftToReturn As Integer

    Friend Sub New(ByVal table As Dictionary(Of Char, String))
        Me.table = table
        Me.bufferIndex = -1
        Me.leftToReturn = -1
    End Sub

    Public Overloads Overrides Function Fallback(ByVal charUnknownHigh As Char, ByVal charUnknownLow As Char, ByVal index As Integer) As Boolean
        ' There's no need to handle surrogates.
        Return False
    End Function

    Public Overloads Overrides Function Fallback(ByVal charUnknown As Char, ByVal index As Integer) As Boolean
        If charUnknown >= ChrW(&H0410) And charUnknown <= ChrW(&H044F) Then
            buffer = table.Item(charUnknown)
            leftToReturn = buffer.Length - 1
            bufferIndex = -1
            Return True
        End If
        Return False
    End Function

    Public Overrides Function GetNextChar() As Char
        Dim charToReturn As Char
        If leftToReturn >= 0 Then
            leftToReturn -= 1
            bufferIndex += 1
            charToReturn = buffer(bufferIndex)
        Else
            charToReturn = ChrW(0)
        End If
        Return charToReturn
    End Function

    Public Overrides Function MovePrevious() As Boolean
        If bufferIndex > 0 Then
            bufferIndex -= 1
            leftToReturn += 1
            Return True
        End If
        Return False
    End Function

    Public Overrides ReadOnly Property Remaining As Integer
        Get
            Return leftToReturn
        End Get
    End Property
End Class
